### PR TITLE
Local cache continued

### LINKED ISSUE #173: Local cache for datasets
In order to avoid hammering a network drive when several processes need access to a dataset at the same time, which can be a problem at MILA, it would be really useful to have an automatic way of caching datasets on a local drive.
Pylearn2 has had that functionality for a while [[1](https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/datasets/cache.py)], and it proved useful to reduce the load on file servers.
One way this could work would be:
- Use a new environment variable, for instance `FUEL_DATA_CACHE` to specify the path of that cache. If that variable is not set, no cache will be used.
- In `fuel.utils.find_in_data_path`, when the original data file is found, try to:
  - see if it is already in the cache, and check that the version is up to date,
  - if not, try to copy if into the cache,
  - if it works, return the path of the cached version, otherwise, return the original path.
- Maybe add an option to `find_in_data_path` to disable caching, for instance `ignore_data_cache`. This can be used by a particular Dataset class to disable caching for its data.

CC @nouiz who worked on the Pylearn2 version.

### PR DESCRIPTION
**Pull request by maximecb**
Taking over the work from #383 (sidenote: you guys should probably close #173 and #383).

All I've done so far is fix the conflicts and get tests running locally. Opening this PR so I can get your feedback as to what remains to be done/fixed.

One thing I did notice is that if I just run `nose2 tests` after installing fuel, I get this error:

```
ConfigurationError: Configuration not set and no default provided: local_data_path
```

Wondering if we should set caching to disabled by default, and what default value `local_data_path` should be set to in order to disable caching, is it just the empty string?

### CODE REVIEW

**Comment by maximecb** [comment]
Following an in-person discussion with @rizar, I will make it so Fuel doesn't give an error if the cache path is not set. I will also add a mention of the caching behavior in the docs.

**Comment by maximecb** [comment]
@rizar @nouiz 

I've made it so caching is disabled by default, and Fuel doesn't complain if the `FUEL_LOCAL_DATA_PATH` variable is not set. I also added a section about caching to the docs. The tests are all passing locally on my machine, with and without caching.

Regarding the failing scrutinizer tests: they are related to some scrutinizer-specific extension of the pep257 linter. I can't run these checks on my machine because their linter seems to be proprietary. Arnaud suggests disabling scrutinizer on this repository, a motion that I would agree with.

Let me know if there is anything else that needs changing.

File: docs/caching.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+point to a shared network drive. For example, when configuring multiple
+machines in a cluster to work on the same data in parallel.
+However, this can easily cause network bandwidth to become saturated.
+
+To avoid this problem, Fuel provides a second configuration variable
+named ``local_data_path``, which can be set in ``~/.fuelrc``. This
+variable points to a filesystem directory to be used to act as a local
+cache for datasets.
+
+This variable can also be set through an environment variable as follows:
```
Comment (rizar) [inline docs/caching.rst]: Can you please mention that caching is currently only implemented in `H5PyDataset`, and that in order to add caching to other dataset one should use `cache_file` function?

**Comment by rizar** [comment]
Regarding the fate of Scrutinizer, it would be great to hear from @bartvm  who started using at the first place.

File: docs/caching.rst (hunk 1 above)
Comment (maximecb) [reply docs/caching.rst]: @rizar Done!

**Comment by maximecb** [comment]
@rizar now passing the scrutiny

**Comment by rizar** [comment]
Thanks!

**Comment by maximecb** [comment]
@rizar NP. Let me know if there are other issues I could potentially help with, on this project or others.

**Pull request merged**